File: advent-of-code-2020/csharp/Day02.cs
using System.Collections.Generic;
using System.IO;
using System.Linq;
using AoCHelper;

namespace advent_of_code_2020.csharp
{
    public sealed class Day02 : BaseDay
    {
        private readonly string[] _input;

        public Day02()
        {
            _input = File.ReadAllLines(InputFilePath).ToArray();
        }

        public static int CountValidPasswords(IEnumerable<string> lines, bool newLogic = false)
        {
            var valid = 0;
            foreach (var line in lines)
            {
                var split = line.Split('-', ' ', ':');
                if (newLogic)
                {
                    var expectedChar = char.Parse(split[2]);
                    var char1 = split[4][int.Parse(split[0]) - 1];
                    var char2 = split[4][int.Parse(split[1]) - 1];
                    if ((char1 == expectedChar || char2 == expectedChar) && char1 != char2) valid += 1;
                }
                else
                {
                    var occurrences = split[4].Count(c => c == char.Parse(split[2]));
                    if (occurrences >= int.Parse(split[0]) && occurrences <= int.Parse(split[1])) valid += 1;
                }
            }

            return valid;
        }

        public override string Solve_1()
        {
            return CountValidPasswords(_input).ToString();
        }

        public override string Solve_2()
        {
            return CountValidPasswords(_input, true).ToString();
        }
    }
}

File: 2020/python/Day02.py
import re

import aoc_utils


def main(parttwo=False):
    count = 0
    for line in aoc_utils.input_string_list():
        line = re.split(r"-|: | ", line)
        occurs = line[-1].count(line[2])
        if (
            parttwo
            and (line[-1][int(line[0]) - 1] + line[-1][int(line[1]) - 1]).count(line[2])
            == 1
        ) or (not parttwo and int(line[0]) <= occurs and int(line[1]) >= occurs):
            count += 1
    return count


print(main())
print(main(True))
